File: natives/src/events/end_file.rs
use crate::events::types::Event;
use crate::types::errors::Error;
use libmpv2_sys::{mpv_end_file_reason, mpv_event};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum EndFileReason {
    EOF,
    Stop,
    Quit,
    Error,
    Redirect,
}

impl From<mpv_end_file_reason> for EndFileReason {
    fn from(value: mpv_end_file_reason) -> Self {
        match value {
            libmpv2_sys::mpv_end_file_reason_MPV_END_FILE_REASON_EOF => Self::EOF,
            libmpv2_sys::mpv_end_file_reason_MPV_END_FILE_REASON_STOP => Self::Stop,
            libmpv2_sys::mpv_end_file_reason_MPV_END_FILE_REASON_QUIT => Self::Quit,
            libmpv2_sys::mpv_end_file_reason_MPV_END_FILE_REASON_ERROR => Self::Error,
            libmpv2_sys::mpv_end_file_reason_MPV_END_FILE_REASON_REDIRECT => Self::Redirect,
            _ => panic!("Unknown end file reason: {}", value),
        }
    }
}

pub(super) unsafe fn event_end_file(raw: &mpv_event) -> Event {
    unsafe {
        let raw_data = raw.data as *mut libmpv2_sys::mpv_event_end_file;
        Event::EndFile {
            reason: EndFileReason::from((*raw_data).reason),
            error: Error::from(raw.error),
            playlist_entry_id: (*raw_data).playlist_entry_id,
            playlist_insert_id: (*raw_data).playlist_insert_id,
            playlist_insert_num_entries: (*raw_data).playlist_insert_num_entries,
        }
    }
}

File: mpv-kt/src/commonMain/kotlin/dev/silenium/mpv/native_bindings/event/EndFile.kt
package dev.silenium.mpv.native_bindings.event

import dev.silenium.libs.foreign.MemorySegment
import dev.silenium.mpv.native_bindings.Error
import dev.silenium.mpv.native_bindings.api.InstantiableLayout
import dev.silenium.mpv.native_bindings.api.NativeEnum
import dev.silenium.mpv.native_bindings.api.NativeStructLayout
import dev.silenium.mpv.native_bindings.api.get

data class EndFile(
    val reason: Reason,
    val error: Error,
    val playlistEntryId: Long,
    val playlistInsertId: Long,
    val playlistInsertNumEntries: Long
): EventData {
    constructor(struct: MemorySegment) : this(
        struct[reason],
        struct[error],
        struct[playlistEntryId],
        struct[playlistInsertId],
        struct[playlistInsertNumEntries],
    )

    enum class Reason(override val value: Int) : NativeEnum<Reason> {
        EOF(0),
        STOP(2),
        QUIT(3),
        ERROR(4),
        REDIRECT(5),
    }

    companion object Layout : NativeStructLayout(), InstantiableLayout<EndFile> {
        val reason = enum<Reason>("reason")
        val error = enum<Error>("error")
        val playlistEntryId = long("playlist_entry_id")
        val playlistInsertId = long("playlist_insert_id")
        val playlistInsertNumEntries = long("playlist_insert_num_entries")

        override fun from(segment: MemorySegment): EndFile = EndFile(segment)
    }
}
